# Conditional Probability

## Introduction: The three card Problem

Suppose one has three cards – one card is blue on both sides, one card is pink on both sides, and one card is blue on one side and pink on the other side. Suppose one chooses a card and place it down showing “blue”. What is the chance that the other side is also blue?

This is an illustration of a famous conditional probability problem. One is given certain information – here the information is that one side of the card is blue – and one wishes to determine the probability that the other side is blue.

Most people think that this probability is $1/2$, but actually this is wrong. The correct answer is demonstrated by simulating this experiment many times. One can do this simulation by hand, but we will illustrate this using a Python script.

Suppose one thinks of this experiments are first choosing a card, and then choosing a side from the card. There are three possible cards, which we call "Blue", "Pink" and "mixed". For the blue card, there are two blue sides; for the pink card, there are two pink sides, and the "mixed" card has a blue side and a pink side.

##### Conditional Probabilities by Simulation

We illustrate using Python to perform this simulation. A data frame `df` with two variables `Card` and `Side` is defined. The `np.random.choice()` function randomly chooses a card and a side by choosing a random row from the data frame. This experiment was repeated 1000 times and the `pd.crosstab()` function is used to classify the outcomes by card and side.

In [1]:
import pandas as pd

df = pd.DataFrame({'Card':['Blue', 'Blue', 'Pink', 'Pink', 'Mixed','Mixed'],
                  'Side':['Blue','Blue', 'Pink','Pink','Blue','Pink']})

In [2]:
df

Card  Side
0   Blue  Blue
1   Blue  Blue
2   Pink  Pink
3   Pink  Pink
4  Mixed  Blue
5  Mixed  Pink

In [3]:
import numpy as np
n_picks = 1000
np.random.seed(1234)

cards = df.iloc[np.random.choice(np.arange(df.shape[0]),size = n_picks, replace = True)]

In [4]:
cards['Card'].value_counts()

Pink     350
Blue     329
Mixed    321
Name: Card, dtype: int64

In [5]:
pd.crosstab(cards.Card, cards.Side)

Side   Blue  Pink
Card             
Blue    329     0
Mixed   155   166
Pink      0   350

One observed side is blue and we are interested in the probability of the event "card is blue". In this experiment, the blue side was observed $329 + 155 = 484$ times – of these, the card was blue 329 times. So the probability the other side is blue is approximated by 329/484 which is close to $2/3$. This example illustrates that one’s intuition can be faulty in figuring out probabilities of the conditional type.

### Selecting Slips of Paper

To illustrate the conditional nature of probabilities, suppose one has a box that has 6 slips of paper – the slips are labeled with the numbers 2, 4, 6, 8, 10, and 12. One selects two slips at random from the box. It is assumed that one is sampling without replacement and the order that one selects the slips is not important. Then one lists all of the possible outcomes. Note that since two numbers are chosen from six, the total number of outcomes will be $^6C_{2} = 15$.

$$S = \{(2, 4), (2, 6), (2, 8), (2, 10), (2, 12), (4, 6), (4, 8), (4, 10), (4, 12) (6, 8), (6, 10), (6, 12), (8, 10), (8, 12), (10, 12)\}$$

Suppose one is interested in the probability the sum of the numbers on the two slips is 14 or higher. Assuming that the 15 outcomes listed above are equally likely, one sees there are 9 outcomes where the sum is 14 or higher and so

$$Prob(\text{sum 14 or higher}) = \frac{9}{15}$$

Next, suppose one is given some new information about this experiment – both of the numbers on the slips are single digits. Given this information, one now has only six possible outcomes. This new sample space is called the reduced sample space based on the new information.

$$ S = \{(2, 4), (2, 6), (2, 8), (4, 6), (4, 8), (6, 8)\} $$

One evaluates the probability $Prob(\text{sum 14 or higher})$ given that both of the slip numbers are single digits. Since there is only one way of obtaining a sum of 14 or higher in our new sample space, one sees

$$Prob(\text{sum 14 or higher}) = \frac{1}{6}$$

**Notation**: Suppose that $E$ is our event of interest and $H$ is our new information. Then one writes the probability of $E$ given the new information $H$ as $Prob(E|H)$, where the vertical line "|" means "conditional on" or given the new information. Here it was found

$$Prob(\text{sum is 14 or higher} | \text{both slip numbers are single digits})$$

How does the probability of "14 or higher" change given the new information? Initially, the probability of 14 and higher was pretty high ($9/15$), but given the new information, the probability dropped to $1/6$. Does this make sense? Yes. If one is told that both numbers are single digits, then one has drawn small numbers and that would tend to make the sum of the digits small.

### Independent Events

One says that events $A$ and $B$ are independent if the knowledge of event $A$ does not change the probability of the event $B$. Using symbols

$$P(A|B) = P(B)$$

### Rolls of two dice

To illustrate the concept of independence, consider an example where one rolls a red die and a white die. Consider the following three events:

- $S = \text{the sum of the two rolls is 7}$
- $E = \text{the red die is an even number}$
- $D = \text{the rolls of the two dice are different}$

##### Are events $S$ and $E$ independent?

1. First one finds the probability one rolls a sum equal to 7, that is, $P(S)$. There are 36 outcomes and 6 outcomes results in a sum of 7, so $P(S) = 6/36$.
2. Next, one finds $P(S | E)$. Given that the red die is an even number (event $E$), note that there are 18 outcomes where $E$ occurs. Of these 18 outcomes, there are 3 outcomes where the sum is equal to 7. So $P(S | E) = 3/18$.
3. Note $P(S | E) = P(S)$, so events $S$ and $E$ are independent. Knowing the red die is even does not change one’s probability of rolling a 7.

##### Are events $S$ and $D$ independent?

To see if these two events are independent, one computes $P(S | D)$ and checks if $P(S | D) = P(S)$. One can show that $P(S | D) = 6/30$. This probability is not equal to $P(S)$ so $S$ and $D$ are not independent events.

##### Conditional Probabilities by Simulation

One can demonstrate conditional probability by the use of the filter() function in the dplyr package. To illustrate, a data frame `df` is constructed with simulated rolls of two dice – the associated variables are `roll_1` and `roll_2`.

In [6]:
import pandas as pd
import numpy as np

np.random.seed(1234)
n_throws = 1000

df = pd.DataFrame({'roll_1': np.random.choice(np.arange(1,7), size = n_throws),
                 'roll_2': np.random.choice(np.arange(1,7), size = n_throws)}) 

We now define a new column `sum` that is the sum of the two rolls.

In [7]:
df['sum'] = df.roll_1 + df.roll_2

In [8]:
df.head()

roll_1  roll_2  sum
0       4       1    5
1       6       1    7
2       5       3    8
3       5       5   10
4       1       2    3

Suppose one is told that the roll of the first die is greater than 3 – how does that information change the probabilities for `sum`? Let's restrict die rolls to only the ones where `roll_1` > 3.

In [9]:
df_gt_3 = df[df['roll_1'] > 3].reset_index(drop = True)

In [10]:
df_gt_3.head()

roll_1  roll_2  sum
0       4       1    5
1       6       1    7
2       5       3    8
3       5       5   10
4       4       1    5

Then the frequencies and corresponding approximate probabilities of different sums are found on these "restricted" die rolls. For example, one sees that the probability $Prob(\text{sum = 10}|\text{roll_1} > 3) \approx 0.177$

In [11]:
df_gt_3[df_gt_3['sum'] ==  10].shape[0]/df_gt_3.shape[0]

0.17693836978131214

In [12]:
df_gt_3['sum'].value_counts()

8     96
10    89
7     78
9     77
6     55
11    45
12    34
5     29
Name: sum, dtype: int64

In [13]:
df_sum = pd.DataFrame(df_gt_3.groupby('sum')['roll_1'].count()).rename(columns = {'roll_1':'count'})

In [14]:
df_sum['probability'] = df_sum['count']/df_sum['count'].sum()

In [15]:
df_sum

count  probability
sum                    
5       29     0.057654
6       55     0.109344
7       78     0.155070
8       96     0.190855
9       77     0.153082
10      89     0.176938
11      45     0.089463
12      34     0.067594

## In Everyday Life

Generally one’s beliefs about uncertain events can change when new information is obtained. Conditional probability provides a way to precisely say how one’s beliefs change. Let’s illustrate this with a simple example.

Suppose one is interested in estimating the population of Philadelphia, Pennsylvania in the current year. Consider three possible events:

- $A = \text{Philadelphia’s population is under one million}$
- $B = \text{Philadelphia’s population is between one and two million}$
- $C = \text{Philadelphia’s population is over two million}$

If one knows little about Philadelphia, then one probably is not very knowledgeable about its population. So initially the probabilities are assigned shown in the next table:

##### Table 3.1 - Probabilities of events about Philadelphia’s population, $P(Event | I)$.
| Event | $$ P(Event | I) $$ |
|:---|:--------|
|Under one million| 0.3|
|between one and two million| 0.3|
|over two million| 0.4|
|Total|1.0|


One is assigning approximately the same probability to each of the three events, indicating that they are all equally likely in his or her mind. These can be viewed as conditional probabilities since they are conditional on one’s initial information – these probabilities are denoted by $P(Event | I)$, where $I$ denotes one’s initial information. Now suppose some new information is provided about Philadelphia’s population. One is not told the current population, but is told that in 1990, Philadelphia was the fifth largest city in the country, and the population of the sixth largest city, San Diego, was 1.1 million in 1990. So this tells one that in 1990, the population of Philadelphia had to exceed 1.1 million. Now one might not be sure about how the population of Philadelphia has changed between 1990 and 2020, but it probably has not changed a significant
amount. So one thinks that

- The population of Philadelphia is most likely to be between 1 and 2 million.
- It is very unlikely that Philadelphia’s population is over 2 million
- There is a small chance that Philadelphia’s population is under 1 million.

One revises his or her probabilities that reflect these beliefs as shown in Table 3.2. These probabilities are denoted as $P(Event | N)$, which are probabilities of these population events conditional on the newer information $N$, in Table 3.2.

##### Table 3.2 - Probabilities of events about Philadelphia’s population, $P(Event | N)$.
| Event | $$P(Event|N)$$ |
|:---|:--------|
|Under one million| 0.2|
|between one and two million| 0.78|
|over two million| 0.02|
|Total|1.0|

Now, additional information is provided. To find the current population of Philadelphia, one looks up the census estimated figures and the population of Philadelphia’s population was reported to be 1,567,872 in 2016. Even though the census number is a few years old, one doesn’t think that the population has changed much – definitely not enough to put in a new category of the table. So one’s probabilities will change again as shown in Table 3.3. We call these probabilities of events conditional on additional information A

##### Table 3.3 - Probabilities of events about Philadelphia’s population, $P(Event | A)$.
| Event |$$P(Event|N)$$|
|:---|:--------|
|Under one million| 0.0|
|between one and two million| 1.0|
|over two million| 0.0|
|Total|1.0|

All of us actually make many judgments every day based on uncertainty. For example, we make decisions about the weather based on information such as the weather report, how it looks outside, and advice from friends. We make decisions about who we think will win a sports event based on what we read in the paper, our knowledge of the teams’ strengths, and discussion with friends. Conditional probability is simply a way of quantifying our beliefs about uncertain events given information.

### In a Two-way table

It can be easier to think about, and compute conditional probabilities when they are found from observed counts in a two-way table.

The following `hs_athlets` data frame containg  high school athletes data in 14 sports that are classified with respect to their sport and their gender. These numbers are recorded in thousands, so the 454 entry in the Baseball/Softball – Male cell means that 454,000 males played baseball or softball this year.

-  Counts of high school athletes by sport and gender.

In [16]:
import pandas as pd

hs_athlets = pd.read_csv('./data/table01.csv', index_col='Sport')
hs_athlets

Male  Female  TOTAL
Sport                                 
Baseball/Softball   454     373    827
Basketball          541     456    997
Cross Country       192     163    355
Football           1048       1   1049
Gymnastics            2      21     23
Golf                163      62    225
Ice Hockey           35       7     42
Lacrosse             50      39     89
Soccer              345     301    646
Swimming             95     141    236
Tennis              145     163    308
Track and Field     550     462   1012
Volleyball           39     397    436
Wrestling           240       4    244
TOTAL              3899    2590   6489

In [17]:
hs_athlets['TOTAL']['TOTAL']

6489

In [18]:
hs_athlets['Male'][:-1].sum()

3899

In [19]:
hs_athlets['Female'][:-1].sum()

2590

Suppose one chooses a high school athlete at random who is involved in one of these 14 sports. Consider several events:

- $F = \text{athlete chosen is female}$
- $S = \text{athlete is a swimmer}$
- $V = \text{athlete plays volleyball}$


What is the probability that the athlete is female? Of the 6489 (thousand) athletes, 2590 were female, so the probability is

In [20]:
round(hs_athlets['Female'][14]/hs_athlets['TOTAL'][14],4)

0.3991

Likewise, the probability that the randomly chosen athlete is a swimmer is

In [21]:
round(hs_athlets.loc['Swimming']['TOTAL']/hs_athlets['TOTAL']['TOTAL'],4)

0.0364

and the probability he or she plays volleyball is

In [22]:
round(hs_athlets.loc['Volleyball']['TOTAL']/hs_athlets['TOTAL']['TOTAL'],4)

0.0672

Next, consider the computation of some conditional probabilities. What is the probability a volleyball player is female? In other words, conditional on the fact that the athlete plays volleyball, what is the chance that the athlete is female:

$$ P(F|V) $$

To find this probability, restrict attention only to the volleyball players in the table.

In [23]:
hs_athlets[hs_athlets.index == 'Volleyball']

Male  Female  TOTAL
Sport                          
Volleyball    39     397    436

Of the 436 (thousand) volleyball players, 397 are female, so

$$P(F|V) = \frac{397}{436} = 0.9106$$

In [24]:
round(hs_athlets[hs_athlets.index == 'Volleyball']['Female']/hs_athlets[hs_athlets.index == 'Volleyball']['TOTAL'],4)

Sport
Volleyball    0.9106
dtype: float64

In [25]:
round(hs_athlets['Female']['Volleyball']/hs_athlets['TOTAL']['Volleyball'],4)

0.9106

What is the probability a woman athlete is a swimmer? In other words, if one knows that the athlete is female, what is the (conditional) probability she is a swimmer, or $P(S|F)$?

Here since one is given the information that the athlete is female, one restricts attention to the "Female" column of counts. There are a total of 2590 (thousand) women who play one of these sports; of these, 141 are swimmers. So

$$P(S|V) = \frac{141}{2590} = 0.0544$$

In [26]:
round(hs_athlets['Female']['Swimming']/hs_athlets['Female']['TOTAL'],4)

0.0544

Are event $F$ and $V$ independent? One can check this several ways. Above it was found that the probability a randomly chosen athlete is a volleyball player is $P(V) = 0.0672$. Suppose one is told that the athlete is a female ($F$). Will that change the probability that she is a volleyball player? Of the 2590 women, 397 are volleyball players, and so $P(V|F)=397/2590 = 0.1533$. Note that $P(V)$ is different from $P(V|F)$ that means that the knowledge the athlete is female has increased one’s probability that the athlete is a volleyball player. So the two events are not independent.

##### Conditional Probabilities in a Two-Way Table in Python

Suppose one has two spinners, each that will record a 1, 2, 3, or 4 with equal probabilities. Suppose the smaller of the two spins is 2 – what is the probability that the larger spin is equal to 4? One can answer this question by use of a simulation experiment. First one constructs a data frame – by two uses of the `sample()` function, 5000 random spins of the first spinner are stored in `Spin_1` and 5000 spins of the second spinner in `Spin_2`

In [27]:
import numpy as np
import pandas as pd

np.random.seed(1234)
n_spins = 5000

df = pd.DataFrame({'Spin_1':np.random.choice(np.arange(1,5), size = n_spins,replace = True),
                   'Spin_2':np.random.choice(np.arange(1,5), size = n_spins,replace = True)})

In [28]:
np.arange(1,5)

array([1, 2, 3, 4])

In [29]:
df['Min'] = np.minimum(df.Spin_1, df.Spin_2)
df['Max'] = np.maximum(df.Spin_1, df.Spin_2)

In [30]:
t1 = df.groupby(['Min','Max']).count()['Spin_1']

In [31]:
t1.unstack()

Max      1      2      3      4
Min                            
1    298.0  606.0  633.0  621.0
2      NaN  321.0  660.0  618.0
3      NaN    NaN  288.0  618.0
4      NaN    NaN    NaN  337.0

Since one is told that the smaller of the two spins is equal to 2, one restricts attention to the row where `Min=2`. One observes that `Max` is equal to 2, 3, 4 with frequencies 321, 660, and 618. So

$$P(\text{Max spin = 4}|\text{Min spin = 2}) = \frac{618}{321+660+618} = 0.386$$

In [32]:
618/(321+660+618)

0.38649155722326456

In [33]:
t1.unstack().loc[2][4]/t1.unstack().loc[2].sum()

0.38649155722326456

## Definition and The Multiplication Rule

In this chapter, conditional probabilities have been computed by considering a reduced sample space. There is a formal definition of conditional probability that is useful in computing probabilities of complicated events.

Suppose one has two events $A$ and $B$ where the probability of event $B$ is positive, that is $P(B) \gt 0$. Then the probability of $A$ given $B$ is defined as the quotient

$$P(A|B) = \frac{P(A \cap B)}{P(B)}$$


##### How many boys?

To illustrate this conditional probability definition, suppose a couple has four children. One is told that this couple has at least one boy. What is the chance that they have exactly two boys?

If one lets $L$ be the event "at least one boy" and $B$ the event "have two boys", one wish to find $P(B|L)$

Suppose one represents the genders of the four children (from youngest to oldest) as a sequence of four letters. For example, the sequence $BBGG$ means that the first two children were boys and the last two were girls. If we represent outcomes this way, there are 16 possible outcomes of four births:

In [34]:
import itertools
a = [''.join(x) for x in itertools.product('BG', repeat=4)]

b = np.matrix([a[0:4],a[4:8],a[8:12],a[12:16]])
pd.DataFrame(b)

0     1     2     3
0  BBBB  BBBG  BBGB  BBGG
1  BGBB  BGBG  BGGB  BGGG
2  GBBB  GBBG  GBGB  GBGG
3  GGBB  GGBG  GGGB  GGGG

In [35]:
len([i for i in a if i.count('B') >= 1])

15

In [36]:
len([i for i in a if i.count('B') == 2])

6

If one assumes that boys and girls are equally likely (is this really true?), then each of the outcomes is equally likely and each outcome is assigned a probability of 1/16. Applying the definition of conditional probability, one has:

$$P(B|L) = \frac{P(B \cap L)}{P(L)}$$

There are 15 outcomes in the set $L$, and 6 outcomes where both events $B$ and $L$ occur. So using the definition

$$P(B|L) = \frac{6/16}{15/16}$$


## The Multiplication Rule

If one takes the conditional probability definition and multiplies both sides of the equation by $P(B)$, one obtains the multiplication rule:

$$ P(A \cap B)=P(B)P(A|B)$$

#### Choosing balls from a random bowl

The multiplication rule is especially useful for experiments that can be divided into stages. Suppose one has two bowls – Bowl 1 is filled with one white and 5 black balls, and Bowl 2 has 4 white and 2 black balls. One first spins the spinner below that determines which bowl to select, and then selects one ball from the bowl. What the chance that the ball one selects is white?

One can demonstrate the multiplication rule by the tree diagram in Figure 3.1. The first set of branches corresponds to the spinner result (choose Bowl 1 or choose Bowl 2) and the second set of branches corresponds to the ball selection.

<img src="./figures/chap_03/figure_3_1.png">

One places numbers on the diagram corresponding to the probabilities that are given in the problem, shown in Figure 3.2.

<img src="./figures/chap_03/figure_3_2.png">

Since one quarter of the spinner region is "Bowl 1", the chance of choosing Bowl 1 is $1/4$ and so the chance of choosing Bowl 2 is $3/4$ – these probabilities are placed at the first set of branches. Also one knows that if Bowl 1 is selected, the chances of choosing a white ball and a black ball are respectively $1/6$ and $5/6$. These conditional probabilities, $P(\text{white}|\text{Bowl 1})$ and $P(\text{black}|\text{Bowl 2})$, are placed at the top set of branches at the second level. Also, if one selects Bowl 2, the conditional probabilities of selecting a white ball and a black ball are given by $P(\text{white}|\text{Bowl 2}) = 4/6$ and $P(\text{black}|\text{Bowl 2}) = 2/6$ – these probabilities are placed at the bottom set of branches.

Now that the probabilities are assigned on the tree, one uses the multiplication rule to compute the probabilities of interest:

- What is the probability of selecting Bowl 1 and selecting a white ball? By the multiplication rule

$$
\begin{align*}
P(\text{Bowl 1} \cap \text{white ball}) &= P(\text{Bowl 1})P(\text{white ball} | \text{Bowl 1}) \\
&= \frac{1}{4} \times \frac{1}{6} = \frac{1}{24}
\end{align*}
$$

One is just multiplying probabilities along the top branch of the tree.

- What is the probability of selecting a white ball? One sees from the tree that there are two ways of selecting a white depending on which bowl is selected. One can either (1) select Bowl 1 and choose a white ball or (2) select Bowl 2 and choose a white ball. One finds the probability of each of the two outcomes and add the probabilities to get the answer.

$$
\begin{align*}
P(\text{white ball}) &= P(\text{Bowl 1} \cap \text{white ball}) + P(\text{Bowl 2} \cap \text{white ball}) \\
&= \frac{1}{4} \times \frac{1}{6} + \frac{3}{4} \times \frac{4}{6} = \frac{13}{24}
\end{align*}
$$

##### Simulating choosing balls from a random bowl

One simulate this balls and bowl experiment on R. Using the `np.random.choice()` function, one simulates 1000 choices of the bowl where the probabilities of choosing Bowl 1 and Bowl 2 are $1/4$ and $3/4$ and places the bowl choices in variable Bowl. In a similar fashion, one simulates 10000 ball selections from Bowl 1 (variable `color_1`) and 10000 selections from Bowl 2 (variable `color_2`). Last, by use of a `np.where`function, one lets the ball color be equal to Color 1 if Bowl 1 is selection, or Color 2 if Bowl 2 is selected

In [37]:
import numpy as np
bowl = np.random.choice([1,2], size=10000, p=[1/4,3/4])

np.random.seed(1234)

color_1 = np.random.choice(['white','black'], size=10000, p=[1/6,5/6])
color_2 = np.random.choice(['white','black'], size=10000, p=[4/6,2/6])

# Color <- ifelse(Bowl == 1, Color_1, Color_2)

a = np.where(bowl == 1, color_1, color_2)

import pandas as pd
df_bowl = pd.DataFrame({'Bowl':bowl,
              'Color':a})

pd.crosstab(index = df_bowl['Bowl'], columns=df_bowl['Color'])
# print(pd.crosstab(index = df_bowl['Bowl'], columns=df_bowl['Color']).iloc[1]['white'])

Color  black  white
Bowl               
1       2079    421
2       2541   4959

The probability that Bowl 1 was selected and a white ball was chosen is approximately equal to 

In [38]:
pd.crosstab(index = df_bowl['Bowl'], columns=df_bowl['Color']).loc[1]['white']/df_bowl.value_counts().sum()

0.0421

 The chance of choosing a white ball is approximated by

In [39]:
pd.crosstab(index = df_bowl['Bowl'], columns=df_bowl['Color'])['white'].sum()/df_bowl.value_counts().sum()

0.538

## The Multiplication Rule Under Independence

When two events $A$ and $B$ are **independent**, then the multiplication rule takes the simple form

$$P(A \cap B) = P(A) \times P(B)$$

Moreover, if one has a sequence of independent events, say $A_1,A_2, \dots, A_k$ then the probability that all events happen simultaneously is the product of the probabilities of the individual events 

$$P(A_1 \cap A_2 \cap \dots \cap A_K) = P(A_1) \times P(A_2) \dots \times P(A_k)$$

By use of the assumption of independent events and multiplying, one finds probabilities of sophisticated events. We illustrate this in several examples.

#### Blood types of Couples  

White Americans have the blood types $0$,$A$,$B$, and $AB$ with respectively proportions 0.45, 0.40, 0.11, and 0.04. Suppose two people in this group are married. 

1. **What is the probability that the man has blood type O and the woman has blood type A?** Let $0_M$ denote the event that the man has $0$ blood type and $A_W$ the event that the woman has $A$ blood type. Since these two people are not related, it is reasonable to assume that $0_M$ and $A_W$ are independent events. Applying the multiplication rule, the probability the couple have these two specific blood types is

$$
\begin{align*} 
P(0_M \cap A_W) &= P(0_M) \times P(A_W)\\ 
&= (0.45) \times (0.40) = 0.18\\ 
\end{align*}
$$

2. **What is the probability the couple have O and A blood types?** This is a different question from the first one since it is not been specified who has the two blood types. Either the man has blood type $0$ and the woman has blood type $A$, or the other way around. So the probability of interest is

$$
\begin{align*} 
P(\text{two have} A,0 \text{ types}) &= P((0_M \cap A_W)) \cup ((0_W \cap A_M)) \\
&= P(0_M \cap A_W) + P(0_W \cap A_M)
\end{align*}
$$

One adds the probabilities since $0_M \cap A_W$ and $0_W \cap A_M$ are different outcomes. One uses the multiplication rule with the independence assumption to find the probability:

$$
\begin{align*} 
P(\text{two have} A,0 \text{ types}) &= P((0_M \cap A_W)) \cup ((0_W \cap A_M)) \\
&= P(0_M \cap A_W) + P(0_W \cap A_M) \\
&= P(0_M) \times P(A_W) + P(0_W) \times P(A_M) \\
&= (0.45) + (0.40) + (0.45) + (0.40) \\
&= 0.36
\end{align*}
$$

3. **What is the probability the man and the woman have the same blood type?** This is a more general question than the earlier parts since one hasn’t specified the blood types – one is just interested in the event that the two people have the same type. There are four possible ways for this to happen: they can both have type $0$, they both have type $A$, they have type $B$, or they have type $AB$. One first finds the probability of each possible outcome and then sum the outcome probabilities to obtain the probability of interest. One obtains

$$
\begin{align*} 
P(\text{same type})  &= P((0_M \cap 0_W) \cup (A_M \cap A_W) \cup (B_M \cap B_W) \cap (AB_M \cap AB_W))\\
&= (0.45)^2 + (0.40)^2 + (0.11)^2 + (0.04)^2 \\
&= 0.3762
\end{align*}
$$

4. **What is the probability the couple have different blood types?** One way of doing this problem is to consider all of the ways to have different blood types – the two people could have blood types $0$ and $A$, types $0$ and $B$, and so on, and add the probabilities of the different outcomes. But it is simpler to note that the event “having different blood types” is the complement of the event “have the same blood type”. Then using the complement property of probability

$$
\begin{align*} 
P(\text{different type})  &= 1 = P(\text{same type})\\
&= 1-P(\text{same type}) \\
&= 1-0.3762 \\
&= 0.6238
\end{align*}
$$

#### A Five-Game Playoff

Suppose two baseball teams play in a "best of five" playoff series, where the first team to win three games wins the series. Suppose the Yankees play the Angels and one believes that the probability the Yankees will win a single game is 0.6. If the results of the games are assumed independent, what is the probability the Yankees win the series?

This is a more sophisticated problem than the first example, since there are numerous outcomes of this series of games. The first thing to note is that the playoff can last three games, four games, or five games. In listing outcomes, one lets $Y$ and $A$ respectively the single-game outcomes "Yankees win" and "Angels win". Then a series result is represented by a sequence of letters. For example, $YYAY$ means that the Yankees won the first two games, the Angels won the third game, and the Yankees won the fourth game and the series. Using this notation, all of the possible outcomes of the five-game series are written below.

|Three games|Four games |Five games|
|:----:|:-----:|:-----:|
| <ins>YYY</ins> | <ins>YYAY</ins>, AAYA |<ins>YYAAY</ins>, AAYYA| 
|AAA| <ins>YAYY</ins>, AYAA |<ins>YAYAY</ins>, AYAYA |
| | <ins>AYYY</ins>, YAAA| <ins>YAAYY</ins> , AYYAA |
| | | <ins>AYYAY</ins>, YAAYA|
| | | <ins>AYAYY</ins>, YAYAA|
| | | <ins>AAYYY</ins>, YYAAA|

One is interested in the probability the Yankees win the series. All of the outcomes above where the Yankees win are underlined. By the assumption of independence, one finds the probability of a specific outcome – for example, the probability of the outcome $YYAY$ as 

$$
\begin{align*}
P(YYAY) &= (0.6) \times (0.6) \times (0.4) \times (0.6) \\
&= 0.0864
\end{align*}
$$

One finds the probability that the Yankees win the series by finding the probabilities of each type of Yankees win and adding the outcome probabilities. The probability of each outcome is written down in the next table.

|Three games|Four games |Five games|
|:----:|:-----:|:-----:|
| P(YYY) = 0.216 | P(YYAY) = 0.0864 |P(YYAAY) = 0.0364| 
| | P(YAYY) = 0.0864 |P(YAYAY) = 0.0364 |
| | P(AYYY) = 0.0864 | P(YAAYY) = 0.0364|
| | | P(AYYAY)= 0.0364|
| | | P(AYAYY) = 0.0364|
| | | P(AAYYY) = 0.0364|

So the probability of interest is given by

$$
\begin{align*}
P(\text{Yankees win series}) &= P(YYY, YYAY,YAYY, \dots) \\
&= 0.216 + 3 \times 0.864 + 6 \times 0.0346 \\
&= 0.683
\end{align*}
$$


In [40]:

s1 = 'YYYA'

[''.join(x) for x in set(itertools.product(s1, repeat = 4))]

['YYYY',
 'YAAY',
 'YYYA',
 'YAAA',
 'AYAA',
 'AAYA',
 'YYAA',
 'YAYA',
 'YYAY',
 'YAYY',
 'AYAY',
 'AAYY',
 'AYYY',
 'AAAY',
 'AYYA',
 'AAAA']

In [41]:
def startwith(word, rep):
    if(word[:rep].count(word[0]) == rep):
        return True
    return False

In [42]:
a = 'AAABB'

startwith(a,3)

True

In [43]:
import itertools

def contains_consecutive(word):
    groups = itertools.groupby(word)
    return([(sum(1 for _ in group)) for label, group in groups])


In [44]:
import itertools

def contains_consecutive(word, rep):
    groups = itertools.groupby(word)
    for label, group in groups:
        if(sum(1 for _ in group) == rep):
            return True
    return False


In [45]:
contains_consecutive('YYYAA',4)

False

In [46]:
s2 = 'YYYAA'
[''.join(x) for x in set(itertools.permutations(s2)) if not startwith(x,3)]

['YAYYA',
 'YYAYA',
 'AYYYA',
 'YAYAY',
 'YAAYY',
 'YYAAY',
 'AYYAY',
 'AYAYY',
 'AAYYY']

In [47]:
[''.join(x) for x in set(itertools.permutations(s2)) if not contains_consecutive(x,3)]

['YAYYA', 'YYAYA', 'YAYAY', 'YAAYY', 'YYAAY', 'AYYAY', 'AYAYY']

In [48]:
r = [''.join(x) for x in set(itertools.permutations(s2))]

In [49]:
r

['YAYYA',
 'YYAYA',
 'YYYAA',
 'AYYYA',
 'YAYAY',
 'YAAYY',
 'YYAAY',
 'AYYAY',
 'AYAYY',
 'AAYYY']

In [50]:
def not_last(word,c,n):
    if(word[:-1].count(c) == n):
        return True
    return False

In [51]:
for i in r:
    print(not_last(i,'Y',3))

True
True
True
True
False
False
False
False
False
False


In [52]:
[''.join(x) for x in set(itertools.permutations(s2)) if not not_last(x,'Y',3)]

['YAYAY', 'YAAYY', 'YYAAY', 'AYYAY', 'AYAYY', 'AAYYY']

In [53]:
[''.join(x) for x in set(itertools.permutations(s2)) if not not_last(x,'A',3) and not contains_consecutive(x,3)]

['YAYYA', 'YYAYA', 'YAYAY', 'YAAYY', 'YYAAY', 'AYYAY', 'AYAYY']

In [54]:
a1 = [''.join(x) for x in itertools.product('AY', repeat=5)]

In [55]:
def correct_sequence(word, n):
    for i in list(set(word)):
        if(word.count(i) > n):
            return False
    return True

In [56]:
 [''.join(x) for x in itertools.product('AY', repeat=5) if ((correct_sequence(x,3) and 
                                                             not startwith(x,3) and
                                                            not not_last(x,'A',3) and
                                                            not not_last(x,'Y',3)))]

['AAYYA',
 'AAYYY',
 'AYAYA',
 'AYAYY',
 'AYYAA',
 'AYYAY',
 'YAAYA',
 'YAAYY',
 'YAYAA',
 'YAYAY',
 'YYAAA',
 'YYAAY']

In [57]:
not_last('AYAAY','A',3)

True

In [58]:
def p_game(games,w, prob):
    #print(len(games)-games.count(w))
    #print((1-prob)**(len(games)-games.count(w)))
    return(prob**games.count(w)*(1-prob)**(len(games)-games.count(w)))

In [59]:
games = "YYAAY"
p = 0.6

p_game(games,'Y',p)

0.03456

In [60]:
3**0

1

In [61]:
[''.join(x) for x in itertools.product('HT', repeat=5)]

['HHHHH',
 'HHHHT',
 'HHHTH',
 'HHHTT',
 'HHTHH',
 'HHTHT',
 'HHTTH',
 'HHTTT',
 'HTHHH',
 'HTHHT',
 'HTHTH',
 'HTHTT',
 'HTTHH',
 'HTTHT',
 'HTTTH',
 'HTTTT',
 'THHHH',
 'THHHT',
 'THHTH',
 'THHTT',
 'THTHH',
 'THTHT',
 'THTTH',
 'THTTT',
 'TTHHH',
 'TTHHT',
 'TTHTH',
 'TTHTT',
 'TTTHH',
 'TTTHT',
 'TTTTH',
 'TTTTT']